sway/config/output.c

```diff
@@ -569,3 +465,137 @@ void free_output_config(struct output_config *oc) {
 	free(oc->background_option);
 	free(oc);
 }
+
+static void handle_swaybg_client_destroy(struct wl_listener *listener,
+		void *data) {
+	wl_list_remove(&config->swaybg_client_destroy.link);
+	wl_list_init(&config->swaybg_client_destroy.link);
+	config->swaybg_client = NULL;
+}
+
+static bool set_cloexec(int fd, bool cloexec) {
+	int flags = fcntl(fd, F_GETFD);
+	if (flags == -1) {
+		sway_log_errno(SWAY_ERROR, "fcntl failed");
+		return false;
+	}
+	if (cloexec) {
+		flags = flags | FD_CLOEXEC;
+	} else {
+		flags = flags & ~FD_CLOEXEC;
+	}
+	if (fcntl(fd, F_SETFD, flags) == -1) {
+		sway_log_errno(SWAY_ERROR, "fcntl failed");
+		return false;
+	}
+	return true;
+}
+
+static bool _spawn_swaybg(char *command) {
+	if (config->swaybg_client != NULL) {
+		wl_client_destroy(config->swaybg_client);
+	}
+	int sockets[2];
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, sockets) != 0) {
+		sway_log_errno(SWAY_ERROR, "socketpair failed");
+		return false;
+	}
+	if (!set_cloexec(sockets[0], true) || !set_cloexec(sockets[1], true)) {
+		return false;
+	}
+
+	config->swaybg_client = wl_client_create(server.wl_display, sockets[0]);
+	if (config->swaybg_client == NULL) {
+		sway_log_errno(SWAY_ERROR, "wl_client_create failed");
+		return false;
+	}
+
+	config->swaybg_client_destroy.notify = handle_swaybg_client_destroy;
+	wl_client_add_destroy_listener(config->swaybg_client,
+		&config->swaybg_client_destroy);
+
+	pid_t pid = fork();
+	if (pid < 0) {
+		sway_log_errno(SWAY_ERROR, "fork failed");
+		return false;
+	} else if (pid == 0) {
+		pid = fork();
+		if (pid < 0) {
+			sway_log_errno(SWAY_ERROR, "fork failed");
+			_exit(EXIT_FAILURE);
+		} else if (pid == 0) {
+			if (!set_cloexec(sockets[1], false)) {
+				_exit(EXIT_FAILURE);
+			}
+
+			char wayland_socket_str[16];
+			snprintf(wayland_socket_str, sizeof(wayland_socket_str),
+				"%d", sockets[1]);
+			setenv("WAYLAND_SOCKET", wayland_socket_str, true);
+
+			execl("/bin/sh", "/bin/sh", "-c", command, NULL);
```

Comment:
Why do we need `sh` here? This requires quoting things, which is not nice

Reply:
What would be the best way of fxing this? The `master` version uses `execvp` with an array of arguments. If I try using that here, it is a VLA which is an error based on the compilation flags.

Reply:
You can use an array with a large-enough size instead of a VLA.

```c
char *cmd[32];
size_t i = 0;

cmd[i++] = "swaybg";
cmd[i++] = "-o";
cmd[i++] = "my-beloved-output";
…

assert(i <= sizeof(cmd) / sizeof(cmd[0]));
```

Reply:
(BTW I'm assuming you'll want to add items in loops and whatnot -- otherwise it would've been simpler)

Reply:
Changed to use `execvp` again. I'm not really sure what the best size for the array is though.

Reply:
We could always dynamically allocate it, but I think a reasonably large value will be enough (e.g. 128).

Reply:
It's at 128 with a comment about how many output configs it will allow for (31 color only, 21 image only, 15 image with fallback color). If anyone complains, we can bump it up or something later

Reply:
Actually, I'll just switch it to dynamic allocation in a few hours (have to go do some stuff now). For some reason, I was thinking it had to be an array with a fixed length...

Reply:
Switched to dynamic allocation